### ISSUE TITLE
Leading spaces

### ISSUE THREAD

**Issue by GoogleCodeExporter** [fracte/keepnote#502]
```
Hey,

first off, KeepNote is an excellent program and I've been looking for something 
like this for a long time. I was also thinking about writing something like 
this myself.

I have to admit there is one small issue which is still a big problem for me 
personally: Spaces at the beginning of a line aren't preserved correctly. This 
is a duplicate issue (see issue 399), but I want to provide a simple solution 
and it would be great if it could be added to KeepNote. Of course I can 
understand if it's not possible - I could also try to build KeepNote on Windows 
myself - but this probably takes more than just a few minutes, so I prefer to 
ask first. :)

================================================================================
Description
================================================================================

I'm not sure if whitespace is stripped here as well, so I will use "[space]" 
instead of the actual space character.

If you make a note ...

Line1
[space]Line2
[space][space]Line3

... and reopen it, it becomes ...

Line1
Line2
[space]Line3

Even worse, if you edit the note again, it becomes ...

Line1
Line2
Line3

The reason is quite simple: "[space][space]" is replaced by "[space]&nbsp;", 
but at the beginning of a line the leading space is automatically ignored later.

I'd like to paste source code into KeepNote and I would also like to indent 
lines with spaces - but both isn't really possible because the formatting would 
change significantly and repeatedly.

================================================================================
Fix
================================================================================

I didn't read through the entire source code, but I think it's the file 
"richtext_html.py" with the line ...

text = text.replace("  ", " &nbsp;")

If we assume that there are only "\n" and "\r\n" line breaks (no "\r" line 
breaks), then it's sufficient to replace the line with these two lines ...

text = text.replace("\n ", "\n&nbsp;")
text = text.replace("  ", " &nbsp;")

There is already an annotation about this in the source code: Maybe the speed 
of the replacements could be improved with one of the following two functions - 
so the text doesn't have to parsed over and over again. Surely there are even 
more efficient ways, but I'd have to test them to be actually sure.

http://docs.python.org/library/cgi.html#cgi.escape
http://docs.python.org/library/xml.sax.utils.html#xml.sax.saxutils.escape

================================================================================
Proposal
================================================================================

This proposal has nothing to do with the problem from above and it also isn't 
very important. :)

In most text editors it's possible to turn (word) wrapping on and off. 
Basically, in KeepNote, this can be achieved by using "No Wrapping", which 
internally uses "<nobr></nobr>".

It would be extremely cool to have an option which turns wrapping on and off - 
but which doesn't change the note itself - just its current "view". I.e. you 
can view the note with and without wrapping by toggling this option. This would 
be very useful for extremely long lines.

Internally, this option could just (temporarily) add "<nobr></nobr>" around the 
entire note (without changing the note itself).

================================================================================

Thanks again for this most excellent program!
```

Original issue reported on code.google.com by `alexr...@gmail.com` on 13 Jul 2011 at 2:21

**Comment by GoogleCodeExporter** [comment]
```
Thanks for this report.  I was able to fix the leading whitespace bug and it 
will be available in keepnote 0.7.5. 
```

Original comment by `matt.rasmus@gmail.com` on 16 Jul 2011 at 5:18
- Changed state: **Started**
- Added labels: **editor**

**Comment by GoogleCodeExporter** [comment]
```
Issue 399 has been merged into this issue.
```

Original comment by `matt.rasmus@gmail.com` on 16 Jul 2011 at 8:21

**Comment by GoogleCodeExporter** [comment]
```
I have now released a testing version of keepnote 0.7.5 that should address 
this issue.  Please try it out and let me know how it works.
http://keepnote.org/#download-test
```

Original comment by `matt.rasmus@gmail.com` on 31 Aug 2011 at 8:40

**Comment by GoogleCodeExporter** [comment]
```
Issue 519 has been merged into this issue.
```

Original comment by `matt.rasmus@gmail.com` on 31 Aug 2011 at 8:41

**Comment by GoogleCodeExporter** [comment]
```
cool, will do, and thx for a GREAT tool!
STG
```

Original comment by `stgr...@gmail.com` on 31 Aug 2011 at 8:42

**Comment by GoogleCodeExporter** [comment]
```
tried it on ten notes with code and it works just fine! :-)
```

Original comment by `stgr...@gmail.com` on 31 Aug 2011 at 8:49

**Comment by GoogleCodeExporter** [comment]
```
Great!  Thanks for the quick reply!
```

Original comment by `matt.rasmus@gmail.com` on 31 Aug 2011 at 9:03

**Comment by GoogleCodeExporter** [comment]
```
:-) no, thank YOU! I always recommend KeepNote in workshops,
hopefully, it'll catch on even more!
```

Original comment by `stgr...@gmail.com` on 31 Aug 2011 at 9:06

**Comment by GoogleCodeExporter** [comment]
```
Not so great here :(

After upgrading to keepnote 0.7.5 (from 0.7.4) I was shown a screen full of 
'New Pages' as per attached screenshot.

Something went definitely wrong here. At first glance, I seems that the 
contents of the page.html files still contain 'useful' (hopefully original) 
information, but it is not being displayed (only the empty gray editor field)
```

Original comment by `epposan@gmail.com` on 1 Sep 2011 at 5:56

Attachments:
- [keepnote-0.7.5-new-page.png](https://storage.googleapis.com/google-code-attachments/keepnote/issue-502/comment-9/keepnote-0.7.5-new-page.png)

**Comment by GoogleCodeExporter** [comment]
```
Following up on comment 9, it seems that (only :-) *all* node.xml files have 
been tampered with (for one, they all contain <attr key="title">New 
Page</attr>, but there's more)

The actual contents in the page.html files are still fine!

I have tried to delete the index.sqlite container and re-index the notebook, 
but to no avail.
```

Original comment by `epposan@gmail.com` on 1 Sep 2011 at 6:36

**Comment by GoogleCodeExporter** [comment]
```
Hmm... There is still an issue with upgrading some notebooks.  I am looking 
into it.  I'll let you know when I have a new testing release.
```

Original comment by `matt.rasmus@gmail.com` on 1 Sep 2011 at 3:13

**Comment by GoogleCodeExporter** [comment]
```
Since the name of the nodes is stored in the node.xml files, I guess there's no 
way to undo this major 'rendering useless' action, other than restoring a 
previous backup, or? 

Maybe you should mention this 'risk' even more explicitly in this case (or even 
withdraw the current version completely) for those who don't have a backup (yet)

Ton.
```

Original comment by `epposan@gmail.com` on 1 Sep 2011 at 3:26

**Comment by GoogleCodeExporter** [comment]
```
Correct, you need to go back to backups to recover from this upgrading error.  
This is why I make testing versions and always tell people to make backups 
first.  Sorry for the inconvenience.
```

Original comment by `matt.rasmus@gmail.com` on 1 Sep 2011 at 3:48

**Comment by GoogleCodeExporter** [comment]
```
I had no such problem, I updated 4 notebooks and all was fine. Just my $0.02 ...
```

Original comment by `stgr...@gmail.com` on 1 Sep 2011 at 4:01

**Comment by GoogleCodeExporter** [comment]
```
I found the problem.  It was an error that occurs only on windows when 
upgrading a notebook.  I have uploaded the fix in-place.  If you have time 
please try out the latest testing version 0.7.5.
http://keepnote.org/#download-test
```

Original comment by `matt.rasmus@gmail.com` on 1 Sep 2011 at 4:09

**Comment by GoogleCodeExporter** [comment]
```
I install the 0.7.5 version under Debian testing amd64. At first look all is OK:

- notebook upgrade without any problems
- spaces at start line are preseved ;-)

Regards
```

Original comment by `li...@slavino.sk` on 1 Sep 2011 at 8:36

**Comment by GoogleCodeExporter** [comment]
```
With this version (0.7.5-110901) the 'destructive tampering' of the node.xml 
files did not occur.

Also, this time there was a database upgrade step from version 5 to version 6, 
which did not occur with the previous 0.7.5-110831 version.


```

Original comment by `epposan@gmail.com` on 1 Sep 2011 at 8:38

**Comment by GoogleCodeExporter** [comment]
```
Hi epposan, thanks for testing it out.  I am again very sorry for the 
corruption you got with the earlier test version.  I try very hard to catch all 
of these destructive bugs before even releasing the first test version, but 
sometimes I miss them (especially if they are windows-specific, since I use 
Linux most).  I try to maintain a high standard for data stability so that 
people feel they can trust keepnote with their notes.  Your testing is very 
helpful.
```

Original comment by `matt.rasmus@gmail.com` on 1 Sep 2011 at 8:51

**Comment by GoogleCodeExporter** [comment]
```
My pleasure! Regards. Ton.
```

Original comment by `epposan@gmail.com` on 1 Sep 2011 at 9:09

**Comment by GoogleCodeExporter** [comment]
```
I noticed that the visible beginning-of-line indentation produced by a leading 
<tab> has changed: it used to produce an indentation worth about 3 characters, 
whereas now it produces an indentation of about 5-½ characters (using sans-9)

before:
-- some text
   text preceded by one <tab>

after:
-- some text
      text preceded by one <tab>

is that intentional?
```

Original comment by `epposan@gmail.com` on 2 Sep 2011 at 6:16

**Comment by GoogleCodeExporter** [comment]
```
This is intentional.  I chose to change the size in order to address issue 190.
```

Original comment by `matt.rasmus@gmail.com` on 2 Sep 2011 at 11:43

**Comment by GoogleCodeExporter** [comment]
Original comment by `matt.rasmus@gmail.com` on 15 Sep 2011 at 2:39
- Changed state: **Done**
